File: bothan-api/client/go-client/grpc.go
package client

import (
	"context"
	"time"

	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials/insecure"

	"github.com/bandprotocol/bothan/bothan-api/client/go-client/proto/bothan/v1"
)

var _ Client = &GrpcClient{}

type GrpcClient struct {
	connection *grpc.ClientConn
	timeout    time.Duration
}

func NewGrpcClient(url string, timeout time.Duration) (*GrpcClient, error) {
	connection, err := grpc.NewClient(url, grpc.WithTransportCredentials(insecure.NewCredentials()))
	if err != nil {
		return nil, err
	}
	return &GrpcClient{connection, timeout}, nil
}

func (c *GrpcClient) GetInfo() (*proto.GetInfoResponse, error) {
	client := proto.NewBothanServiceClient(c.connection)
	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
	defer cancel()

	return client.GetInfo(ctx, &proto.GetInfoRequest{})
}

func (c *GrpcClient) UpdateRegistry(ipfsHash string, version string) error {
	client := proto.NewBothanServiceClient(c.connection)
	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
	defer cancel()

	_, err := client.UpdateRegistry(ctx, &proto.UpdateRegistryRequest{IpfsHash: ipfsHash, Version: version})
	return err
}

func (c *GrpcClient) PushMonitoringRecords(uuid, txHash string) error {
	client := proto.NewBothanServiceClient(c.connection)
	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
	defer cancel()

	_, err := client.PushMonitoringRecords(ctx, &proto.PushMonitoringRecordsRequest{Uuid: uuid, TxHash: txHash})
	return err
}

func (c *GrpcClient) GetPrices(signalIDs []string) (*proto.GetPricesResponse, error) {
	// Create a client instance using the connection.
	client := proto.NewBothanServiceClient(c.connection)
	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
	defer cancel()

	return client.GetPrices(ctx, &proto.GetPricesRequest{SignalIds: signalIDs})
}

File: bothan-api/client/rust-client/src/client/grpc.rs
//! Bothan API Rust gRPC client implementation.
//!
//! Provides async gRPC client for Bothan API Server.

use std::str::FromStr;
use std::sync::Arc;

use tokio::sync::Mutex;
use tonic::transport::{Channel, Endpoint};
use tonic::{Request, Status};

use crate::proto::bothan::v1::{
    BothanServiceClient, GetInfoRequest, GetInfoResponse, GetPricesRequest, GetPricesResponse,
    PushMonitoringRecordsRequest, UpdateRegistryRequest,
};

pub struct GrpcClient {
    client: Arc<Mutex<BothanServiceClient<Channel>>>,
}

/// Async gRPC client for Bothan API Server.
impl GrpcClient {
    pub fn new(client: BothanServiceClient<Channel>) -> Self {
        Self {
            client: Arc::new(Mutex::new(client)),
        }
    }

    pub async fn connect(addr: &str) -> Result<Self, tonic::transport::Error> {
        let endpoint = Endpoint::from_str(addr)?;
        let client = Arc::new(Mutex::new(BothanServiceClient::connect(endpoint).await?));
        Ok(GrpcClient { client })
    }

    pub async fn get_info(&self) -> Result<GetInfoResponse, Status> {
        let get_info_request = GetInfoRequest {};
        let request = Request::new(get_info_request);
        let response = self.client.lock().await.get_info(request).await?;
        Ok(response.into_inner())
    }

    pub async fn update_registry(&self, ipfs_hash: &str, version: &str) -> Result<(), Status> {
        let update_registry_request = UpdateRegistryRequest {
            ipfs_hash: ipfs_hash.into(),
            version: version.into(),
        };
        let request = Request::new(update_registry_request);
        let _ = self.client.lock().await.update_registry(request).await?;
        Ok(())
    }

    pub async fn push_monitoring_records(
        &self,
        uuid: &str,
        tx_hash: &str,
        signal_ids: &[&str],
    ) -> Result<(), Status> {
        let push_monitoring_records_request = PushMonitoringRecordsRequest {
            uuid: uuid.into(),
            tx_hash: tx_hash.into(),
            signal_ids: signal_ids.iter().map(|s| s.to_string()).collect(),
        };
        let request = Request::new(push_monitoring_records_request);
        let _ = self
            .client
            .lock()
            .await
            .push_monitoring_records(request)
            .await?;
        Ok(())
    }

    pub async fn get_prices(&self, signal_ids: &[&str]) -> Result<GetPricesResponse, Status> {
        let get_prices_request = GetPricesRequest {
            signal_ids: signal_ids.iter().map(|s| s.to_string()).collect(),
        };
        let request = Request::new(get_prices_request);
        let response = self.client.lock().await.get_prices(request).await?;
        Ok(response.into_inner())
    }
}
